### PR TITLE
Utils for Metric & Logging

### PR DESCRIPTION
**Pull request by ajy8456**
Various utils class & method for computing metric and write logging.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -11,3 +11,5 @@ torch==1.7.1+cu101
 # torch-xla is a bit of a mess. configure as needed on your machine.
 https://storage.googleapis.com/tpu-pytorch/wheels/torch_xla-1.7-cp36-cp36m-linux_x86_64.whl
 torchvision==0.8.2+cu101
+tensorboard
+cv2==4.5.1
```
Comment (yycho0108) [inline requirements.txt:15]: We're not using `requirements.txt` anymore, but the correct package name is `opencv-python`
```suggestion
opencv-python==4.5.1
```

File: src/top/run/box_generator.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    def from_transformation(cls, rotation, translation, scale):
+        """Constructs an oriented bounding box from transformation and scale."""
+        if rotation.size != 3 and rotation.size != 9:
+            raise ValueError('Unsupported rotation, only 3x1 euler angles or 3x3 '
+                             + 'rotation matrices are supported. ' + rotation)
+        if rotation.size == 3:
+            rotation = Rotation.from_rotvec(rotation.tolist()).as_dcm()
+        scaled_identity_box = cls.scaled_axis_aligned_vertices(scale)
+        vertices = np.zeros((NUM_KEYPOINTS, 3))
+        for i in range(NUM_KEYPOINTS):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:64]: For loop is not needed here if you think about which components of the tensor interact with each other.
We have a rotation matrix of shape (3,3) and an input list of vertices (N,3) that is transformed to the output list of vertices (N,3).

File: src/top/run/box_generator.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+    def __repr__(self):
+        representation = 'Box: '
+        for i in range(NUM_KEYPOINTS):
+            representation += '[{0}: {1}, {2}, {3}]'.format(i, self.vertices[i, 0],
+                                                               self.vertices[i, 1],
+                                                               self.vertices[i, 2])
+        return representation
+
+    def __len__(self):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:76]: What does a `length` of a box even mean? I think we should remove this function in favor of a more explicit `num_keypoints()` function or property.

File: src/top/run/box_generator.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        if transformation.shape != (4, 4):
+            raise ValueError('Transformation should be a 4x4 matrix.')
+
+        new_rotation = np.matmul(transformation[:3, :3], self.rotation)
+        new_translation = transformation[:3, 3] + (np.matmul(transformation[:3, :3], self.translation))
+        
+        return Box.from_transformation(new_rotation, new_translation, self.scale)
+
+    @classmethod
+    def scaled_axis_aligned_vertices(cls, scale):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:103]: Do we really need this method? Why don't we just use `UNIT_BOX * scale`?

File: src/top/run/box_generator.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                scale[axis] += np.linalg.norm(vertices[begin, :] - vertices[end, :])
+            scale[axis] /= 4.
+
+        x = cls.scaled_axis_aligned_vertices(scale)
+        system = np.concatenate((x, np.ones((NUM_KEYPOINTS, 1))), axis=1)
+        solution, _, _, _ = lstsq(system, vertices, rcond=None)
+        orientation = solution[:3, :3].T
+        translation = solution[3, :3]
+        
+        return orientation, translation, scale
```
Comment (yycho0108) [inline src/top/run/box_generator.py:151]: It feels a bit unnatural to me that `Box.fit(...)` returns a tuple of `(orientation,translation,scale)` instead of an instance of `Box`.

File: src/top/run/box_generator.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+    def inside(self, point):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:153]: I think `Box.inside(point)` is a confusing (and inverted) name. `Box.contains(point)` is more accurate.

```suggestion
    def contains(self, point):
```

File: src/top/run/box_generator.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        Tests whether a given point is inside the box.
+        Brings the 3D point into the local coordinate of the box.
+        In the local coordinate, the looks like an axis-aligned bounding box.
+        Next checks if the box contains the point.
+        Args:
+            point: A 3*1 numpy vector.
+        Returns:
+            True if the point is inside the box, False otherwise.
+        """
+        inv_trans = np.linalg.inv(self.transformation)
```
Comment (yycho0108) [inline src/top/run/box_generator.py:164]: SE3 transformation does not need an explicit linear inverse, it would be preferable to avoid expensive `np.linalg.inv`. Also, if `inside()` is going to be called often, we should prefer to cache `inv_trans`.

File: src/top/run/box_generator.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        scale = self.scale
+        point_w = np.matmul(inv_trans[:3, :3], point) + inv_trans[:3, 3]
+        for i in range(3):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:167]: Prefer to vectorize: `np.any(...)` instead of `for i in range(3)`

File: src/top/run/box_generator.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        """
+        Compute the volume of the parallelpiped or the box.
+        For the boxes, this is equivalent to np.prod(self.scale).
+        However for parallelpiped, this is more involved.
+        Viewing the box as a linear function we can estimate the volume using a determinant.
+        This is equivalent to sp.ConvexHull(self._vertices).volume
+        Returns:
+            volume (float)
+        """
+        if self._volume is None:
```
Comment (yycho0108) [inline src/top/run/box_generator.py:211]: Why do we need to do anything fancy when we can just do `np.prod(self.scale)`?

File: src/top/run/box_generator.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+            i = self._vertices[2, :] - self._vertices[1, :]
+            j = self._vertices[3, :] - self._vertices[1, :]
+            k = self._vertices[5, :] - self._vertices[1, :]
+            sys = np.array([i, j, k])
+            self._volume = abs(np.linalg.det(sys))
+        return self._volume
+
+    @property
+    def transformation(self):
```
Comment (yycho0108) [inline src/top/run/box_generator.py:220]: Keeping around both `_transformation` and `_rotation,_translation` feels like a recipe for error (inconsistency) to me. One of them should be a derived property of another.

Comment (yycho0108) [reply src/top/run/box_generator.py:220]: We could instead prefer:
```
@property
def rotation(self):
    # ( check if _transformation exists here, if you'd like )
    return self._transformation[:3,:3]
```

File: src/top/run/draw_bounding_box.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
+"""
+Methods for drawing a bounding box on an image.
+Reference: https://github.com/google-research-datasets/Objectron/blob/master/objectron/dataset/graphics.py
+"""
+
+import cv2
+import numpy as np
+import src.top.run.box_generator as Box
```
Comment (yycho0108) [inline src/top/run/draw_bounding_box.py:8]: Why are we importing `box_generator` as `Box` instead of just calling it `box_generator`? We also have a class named `Box` which can be *very* confusing.

File: src/top/run/draw_bounding_box.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+
+_LINE_TICKNESS = 10
```
Comment (yycho0108) [inline src/top/run/draw_bounding_box.py:11]: nit: typo
```suggestion
_LINE_THICKNESS = 10
```

File: src/top/run/loss_util.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    feat = feat.view(feat.size(0), -1, feat.size(3))
+    feat = _gather_feat(feat, ind)
+    return feat
+
+
+def _neg_loss(pred, gt):
+    """
+    Focal loss same as CornerNet.
+    Args:
+        pred (batch x c x h x w)
```
Comment (yycho0108) [inline src/top/run/loss_util.py:32]: we should stick to either `(batch x channels x height x width)` or `(b x c x h x w)`. Personally I prefer `BCHW` or `(B,C,H,W)` since `x`s can be quite confusing. `NCHW` is also quite common.

File: src/top/run/loss_util.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+def _transpose_and_gather_feat(feat, ind):
+    feat = feat.permute(0, 2, 3, 1).contiguous()
+    feat = feat.view(feat.size(0), -1, feat.size(3))
+    feat = _gather_feat(feat, ind)
+    return feat
+
+
+def _neg_loss(pred, gt):
+    """
+    Focal loss same as CornerNet.
```
Comment (yycho0108) [inline src/top/run/loss_util.py:30]: But why is it called `_neg_loss` if this is `Focal loss`?

File: src/top/run/loss_util.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            gt_regr (batch x max_objects x dim)
+            mask (batch x max_objects)
+    """
+    num = mask.float().sum()
+    mask = mask.unsqueeze(2).expand_as(gt_regr).float()
+
+    regr = regr * mask
+    gt_regr = gt_regr * mask
+    
+    regr_loss = nn.functional.smooth_l1_loss(regr, gt_regr, size_average=False)
```
Comment (yycho0108) [inline src/top/run/loss_util.py:70]: This is up to you but I think it's common to abbreviate `nn.functional` as `F`.

File: src/top/run/loss_util.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self):
+        super(FocalLoss, self).__init__()
+        self.neg_loss = _neg_loss
+
+    def forward(self, out, target):
+        return self.neg_loss(out, target)
+
+class RegLoss(nn.Module):
+    """
+    Regression loss for an output tensor
```
Comment (yycho0108) [inline src/top/run/loss_util.py:87]: What does `output tensor` mean here?

File: src/top/run/metrics_util.py [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+        self.hit = np.zeros(self.size, dtype=np.float)
+        self.miss = np.zeros(self.size, dtype=np.float)
+
+    def reset(self):
+        self.hit = np.zeros(self.size, dtype=np.float)
+        self.miss = np.zeros(self.size, dtype=np.float)
+
+    def record_hit_miss(self, metric, greater=True):
+        """Records the hit or miss for the object based on the metric threshold."""
+        # NOTE(Jiyong): greater True -> find greater than threshold / False -> find worse than threshold
```
Comment (yycho0108) [inline src/top/run/metrics_util.py:26]: This is a comment about the argument - as such, it should be included in the function docstring.

File: src/top/run/metrics_util.py [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+
+        for i in range(self.size):
```
Comment (yycho0108) [inline src/top/run/metrics_util.py:28]: Can we vectorize this loop? e.g.
```
self.hit += (metric >= self.thresholds) if greater else (metric <= self.thresholds)
```

File: src/top/run/metrics_util.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+
+    def add_error(self, error):
+        """Adds an error."""
+        self._errors.append(error)
+
+    def compute_accuracy(self, thresh=0.1):
+        """Computes accuracy for a given threshold."""
+        if not self._errors:
+            return 0
+        return len(np.where(np.array(self._errors) <= thresh)[0]) * 100. / len(self._errors)
```
Comment (yycho0108) [inline src/top/run/metrics_util.py:52]: If we're going to use percentages this choice should be documented. Since we're multiplying by 100 I'm assuming we're using percentages here.

File: src/top/run/metrics_util.py [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+class IoU(object):
+    """General Intersection Over Union cost for Oriented 3D bounding boxes."""
+
+    def __init__(self, box1, box2):
+        self._box1 = box1
+        self._box2 = box2
+        self._intersection_points = []
+
+    def iou(self):
+        """Computes the exact IoU using Sutherland-Hodgman algorithm."""
```
Comment (yycho0108) [inline src/top/run/metrics_util.py:139]: docstring is not true.

File: src/top/run/metrics_util.py [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+            self.aps[i] = self.compute_ap(recall, precision)
+
+
+# Global variables for IoU
+_PLANE_THICKNESS_EPSILON = 0.000001
+_POINT_IN_FRONT_OF_PLANE = 1
+_POINT_ON_PLANE = 0
+_POINT_BEHIND_PLANE = -1
+
+class IoU(object):
```
Comment (yycho0108) [inline src/top/run/metrics_util.py:130]: Great!

File: src/top/run/train_util.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python3
```
Comment (yycho0108) [inline src/top/run/train_util.py:1]: I think this file would go away if you rebase from `main`.
```
git fetch origin
git checkout logging
git rebase origin/main
```

**Review by yycho0108 — APPROVED** [review]
Overall LGTM, let's try to clean things up a bit.
